cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -259,25 +289,93 @@ public void visitEnd() {
               entityPart = new EntityPartDetail(fieldName, true);
             } else {
               // preference to method parameters
-              if (paramAnnotation.containsKey(name)) {
+              if (paramPositions.containsKey(name)) {
                 entityPart = new EntityPartDetail(name, false);
               } else if (fieldDetails.containsKey(name)) {
                 entityPart = new EntityPartDetail(name, true);
               } else {
                 // Didn't find a named method parameter or class field for the given entity name
                 throw new IllegalArgumentException(String.format("No named method parameter or a class field found " +
-                                                                   "with name %s in class %s for method %s whereas " +
-                                                                   "it was specified in %s annotation", name,
-                                                                 className, methodName,
-                                                                 AuthEnforce.class.getSimpleName()));
+                                "with name '%s' in class '%s' for method '%s' whereas " +
+                                "it was specified in '%s' annotation", name,
+                        className, methodName,
+                        AuthEnforce.class.getSimpleName()));
               }
             }
             entityPartDetails.add(entityPart);
           }
+          // verify that the entity parts specified in annotation is valid to do enforcement on the given enforceOn
+          verifyEntityParts(entityPartDetails, nodeProcessor.getEnforceOn(), parameterDetails, paramPositions);
           // Store all the information properly for the second pass
           methodAnnotations.put(new Method(methodName, methodDesc),
                                 new AnnotationDetail(entityPartDetails, nodeProcessor.getEnforceOn(),
-                                                     nodeProcessor.getActions(), paramAnnotation));
+                                                     nodeProcessor.getActions(), paramPositions));
+        }
+
+        /** Helper Methods **/
+
+        private void verifyEntityParts(List<EntityPartDetail> entityPartDetails, Type enforceOn,
+                                       Map<Integer, ParameterDetail> parameterAnnotationNode,
+                                       Map<String, Integer> paramAnnotation) {
+          Type entityPartType = getEntityPartType(parameterAnnotationNode, paramAnnotation, entityPartDetails.get(0));
+          // TODO: Later change this to even work with parent type as we will support enforceOn parent type
```

Comment:
I move the TODO in line below the comment. Currently we don't support enforcing on parent EntityId while providing a named entityId on which enforcement will be done. For example you cannot do the following at this point

```
  @AuthEnforce(entities = "applicationId", enforceOn = NamespaceId.class, actions = Action.ADMIN)
  public synchronized void someOp(@Name("applicationId") final ApplicationId application) throws Exception {
```

Although this is what we want to support. This TODO says that we should allow this in future.